### ISSUE TITLE
Investigate alternative language detection libraries

### ISSUE THREAD

**Issue by ryanslade** [sourcegraph/sourcegraph#6713]
Sourcegraph currently use https://github.com/src-d/enry

It's worth looking into https://github.com/boyter/scc or other alternatives as they might be faster.  An implementation that allows streaming rather than taking a byte slice would simplify the code too.

**Comment by slimsag** [comment]
Exposing the COCOMO calculation from scc would also be quite cool.

**Comment by unknwon** [comment]
If I'm not mistaken, the only drawback is that boyter/scc has to be executed in a working tree?

**Comment by rvantonder** [comment]
> If I'm not mistaken, the only drawback is that boyter/scc has to be executed in a working tree?

@boyter :wave: :-) care to comment?

**Comment by slimsag** [comment]
boyter/scc has a full Go API, I don't see why we would need to invoke it in a working tree? It seems we can just give it the file contents as we would for other packages: https://github.com/boyter/scc#api-support

**Comment by boyter** [comment]
Yep that's correct. You can run it against byte content if you want using the API. A few other tools do this such as https://github.com/pinpt/ripsrc which might be a good example.

You can also use `scc` to detect the language for you if you wish which appeared to be the main question. For example here is how I am doing it in a fork of gitea with code processing afterwards for counts as well. Included below an extract.

```
languages, _ := processor.DetectLanguage(blob.Name())
language := processor.DetermineLanguage(blob.Name(), "Plain Text", languages, []byte(fileContent))

fileJob := processor.FileJob{
	Content: []byte(fileContent),
	Language: language,
}
processor.ProcessConstants() // Required to load the language information and need only be done once
processor.CountStats(&fileJob)
```

The reason for the 2 step detection is to deal with languages which share extensions such as Coq and Verilog.

Let me know if that helps. I need to improve the documentation for the API so knowing what pain points there would assist.

**Comment by keegancsmith** [comment]
We have been having perf issues with the current library. This is likely more to do with having to read in the whole repo than the library. But maybe investigating an alternative library would help us avoid needing to read in most files (and mostly rely on file names)
